onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -430,14 +464,17 @@ Status CUDAExecutionProvider::OnRunEnd(bool sync_stream, const onnxruntime::RunO
 }
 
 bool CUDAExecutionProvider::IsGraphCaptureEnabled() const {
-  return info_.enable_cuda_graph;
+  return info_.enable_cuda_graph && GetPerThreadContext().IsGraphCaptureAllowedOnRun();
 }
 
-bool CUDAExecutionProvider::IsGraphCaptured() const {
+bool CUDAExecutionProvider::IsGraphCaptured(int graph_annotation_id) const {
+  ORT_ENFORCE(GetPerThreadContext().SetCudaGraphAnnotationId(graph_annotation_id) == Status::OK());
   return GetPerThreadContext().IsGraphCaptured();
```

Comment:
Shall we change interface like GetPerThreadContext().IsGraphCaptured(graph_annotation_id)?